FsAutoComplete/CompilerServiceInterface.fs

```diff
@@ -0,0 +1,159 @@
+namespace FsAutoComplete
+
+open System
+open System.IO
+open Microsoft.FSharp.Compiler
+open Microsoft.FSharp.Compiler.SourceCodeServices
+
+type ParseAndCheckResults(parse: FSharpParseFileResults,
+                          check: FSharpCheckFileResults,
+                          version: int) =
+  let parseResults = parse
+  let checkResults = check
+  let _version = version
+
+  member x.TryGetMethodOverrides (lines: string[]) (line: int) (col: int) =
+    // Find the starting point, ideally right after the first '('
+    let lineCutoff = line - 3
+    let commas, line, col =
+      let rec prevPos (line,col) =
+        match line, col with
+        | 1, 1
+        | _ when line < lineCutoff -> 1, 1
+        | _, 1 ->
+           let prevLine = lines.[line - 2]
+           if prevLine.Length = 0 then prevPos(line-1, 1)
+           else line - 1, prevLine.Length
+        | _    -> line, col - 1
+
+      let rec loop commas depth (line, col) =
+        if (line,col) <= (1,1) then (0, line, col) else
+        let ch = lines.[line - 1].[col - 1]
+        let commas = if depth = 0 && ch = ',' then commas + 1 else commas
+        if (ch = '(' || ch = '{' || ch = '[') && depth > 0 then loop commas (depth - 1) (prevPos (line,col))
+        elif ch = ')' || ch = '}' || ch = ']' then loop commas (depth + 1) (prevPos (line,col))
+        elif ch = '(' || ch = '<' then commas, line, col
+        else loop commas depth (prevPos (line,col))
+      match loop 0 0 (prevPos(line,col)) with
```

Comment:
Yes, and yes I agree.

Using the tokenizer would be nice, but tokenizing backwards without maintaining the state of the whole file can also be fragile. Have you switched over to that approach completely? Is there anything in the state apart from the the `#if` stack that we need to worry about? I suppose there are also multiline strings. I really feel that there should be canonical implementations of this, and the ident cracking, in FSharp.Compiler.Service, would you agree?